### DISCUSSION TITLE
UUID Relationship between Towny Banks and Essentials Economy

### DISCUSSION THREAD

**Discussion by civicraft** [TownyAdvanced/Towny#6217] [Q&A] [answered]
Hi there, apologies if this isn't the correct place to discuss this.

I am currently setting up a towny server and have a goal of creating a dynamic website that displays a list of towns/nations and displays a profile page about them.

By using SQL Database I can easily access most of the information I need... such as Mayor Names, Town Names, Nation Names, Found Dates etc, but obviously the Town Bank, Nation Bank and Player Balances are not stored in the Towny Database, instead (in my case) they are stored in .yml files in EssentialsX's Economy.

When trying to find the balance of a player, it's easy, as the Towny Database lists the UUID of a player, and the corresponding EssentialsX file is named using this UUID.

For example... If a player had a UUID of "0001" in the Towny Database, the corresponding Essentials economy file would be named "0001.yml", which makes it easy to read the YML file and retrieve a players balance.

The same cannot be said when it comes to towns and nations...

I have created a town named Test Town and a nation named Test Nation.
Unlike players, the towns and nations do not have matching UUIDs in Towny Database and Essentials

**Test Player:**
Towny UUID: fd01a953-3b5f-4a88-8580-af480bd24780
Essentials UUID/YML: fd01a953-3b5f-4a88-8580-af480bd24780.yml

**Test Town:**
Towny UUID: 394a3ebe-6521-4965-8d50-45cd942e8bc0
Essentials UUID/YML: 661ff52e-a323-320e-b4c5-6d9defc77fc3.yml

My question is... if the UUID don't match, how are towny and essentials accessing the correct files?

My only possible thought is that Essentials is taking the town name, which would be "test_town" and then searching each .yml file for an NPC named "PREFIX_test_town" which would seem a long process as opposed to just matching UUIDs.

So atm, I personally don't have a reliable way to retrieve town/nation balances to display on the dynamic site, as I do with players. (apart from implementing a script that trudges through each .yml file retrieving balances and assigning them to a table)

Any nudges in the right direction towards understanding the relationship between the mis-matched UUIDS would be very grateful.

My long term plan would be to release the dynamic site (probably in .php/mysql form) as open source so that other towny servers could have player/town/nation dynamic profiles.

Thanks in advance for any help/discussion based on this aspect of towny =)

**Comment by LlmDl** [comment, marked as answer]
You've stumbled onto an old issue that stems from Vault not using UUIDs.

Towny makes up a UUID for a town. Towny makes a town bank account in Essentials using the String town name (because we cannot do this via UUID.) During the account creation in Essentials, they assign it a UUID (because everyone uses UUIDs for things now,) and it of course ends up being different.

I have a solution for this issue already but it has not been accepted yet: https://github.com/MilkBowl/VaultAPI/pull/138

**Reply by civicraft** [reply]
Thanks LlmDl, I appreciate the response.

I read through your solution, it seems like a perfect fix. Hopefully it'll be accepted in future.

I can still put together dynamic profiles thanks to how you've set up the towny databse, so I will just omit balances for the moment.

I'm also hoping to put together a simple sql databse for histories, as a lot of towny servers create wikis. 

Something simple, that when a town is created or falls, an entry is created in a table such as "towny_historical_events" with the town name, relevant dates and a flag such as "town_created", and then a simple page that creates a table from entries with relevant flags.

Maybe if I manage to get the hang of github I'll be able to send them over once done for integration consideration.

Thanks for all your work, my family has enjoyed countless hours of towny over the years.
